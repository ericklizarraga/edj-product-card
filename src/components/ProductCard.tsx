
import styles from '../styles/styles.module.css';

import React, { createContext } from 'react';
import { useProduct } from '../hooks/useProduct';
import { ProductContextProps, Product, onChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/product.inteface';


export interface Props {
    product: Product;
    // children?: ReactElement | ReactElement[];
    children?: (handlers: ProductCardHandlers) => React.JSX.Element;
    className?: string;
    style?: React.CSSProperties;
    onChange?: (args: onChangeArgs) => void;
    value?: number;
    inicialValues?: InitialValues;
}


export const ProductContext = createContext<ProductContextProps>({} as ProductContextProps);



export const ProductCard = function producto({ children, product, className, style, onChange, value, inicialValues }: Props) {



    const { counter, increaseBy, maxCount,reset,isMaxCountReached } = useProduct({ product, onChange, value, inicialValues });



    return (
        <ProductContext.Provider
            value={{
                product,
                useProduct: {
                    counter,
                    increaseBy,
                    maxCount
                }
            }}
        >
            <div
                className={`${styles.productCard} ${className}`}
                style={style}
            >
                {
                !!children 
                    &&  children({
                            count:counter,
                            increaseBy,
                            maxCount,
                            isMaxCountReached,
                            product,
                            reset,
                        })
                }
            </div>
        </ProductContext.Provider>
    )
}
