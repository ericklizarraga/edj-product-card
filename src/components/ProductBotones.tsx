
import React from "react";
import { useCallback, useContext } from 'react';
import styles from '../styles/styles.module.css';
import { ProductContextProps } from '../interfaces/product.inteface';
import { ProductContext } from './ProductCard';

export interface Props {
    className?: string;
    style?:React.CSSProperties;
}

export const ProductBotons = ({ className,style }: Props) => {

    const { useProduct: { increaseBy, counter,maxCount } } = useContext<ProductContextProps>(ProductContext);

    const isMaxReached = useCallback(()=>{
        return counter === maxCount;
    },[counter,maxCount])();

    console.log({counter,maxCount,isMaxReached});
    
    return (
        <div 
            className={`${styles.buttonsContainer} ${className}`}
            style={style}
        >
            <button className={styles.buttonMinus} onClick={() => increaseBy(-1)} > - </button>

            <div className={styles.countLabel}>  {counter} </div>

            <button 
                className={`${styles.buttonAdd} ${ !!isMaxReached ? styles.disabled: null}`} 
                onClick={() => increaseBy(1)}
            > + </button>
        </div>
    );
}
