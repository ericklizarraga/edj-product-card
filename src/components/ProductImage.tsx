import React from "react";
import  { useContext } from "react";
import { ProductContextProps } from "../interfaces/product.inteface";
import { ProductContext } from "./ProductCard";
import styles from '../styles/styles.module.css';
import noImage from '../assets/no-image.jpg';


export interface Props  {
    img?:string;
    className?:string;
    style?:React.CSSProperties;
}

export const ProductImage = ({ img = '',className,style }:Props) => {

    const { product } = useContext<ProductContextProps>(ProductContext);

    if (!img) {
        img = product.img as string;
    }

    return (
        <img 
            className={`${styles.productImg} ${className}`} 
            src={img ? img : noImage} alt='producto img'
            style={style}
         />
    )
}