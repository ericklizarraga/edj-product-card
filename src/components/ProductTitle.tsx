import React from "react";
import { useContext } from "react";
import { ProductContextProps } from "../interfaces/product.inteface";
import { ProductContext } from "./ProductCard";
import styles from '../styles/styles.module.css';

export interface Props {
    title?: String;
    className?:string;
    style?:React.CSSProperties;
}

export const ProductTitle = ({ title,className,style }: Props) => {

    const { product } = useContext<ProductContextProps>(ProductContext);

    if (!title) {
        title = product.title;
    }

    return (
        <span 
            className={`${styles.productDescription} ${className}`}
            style={style}
        >{title}</span>
    );
}