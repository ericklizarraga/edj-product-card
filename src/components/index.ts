export { ProductImage } from './ProductImage';
export { ProductTitle } from './ProductTitle';
export { ProductBotons } from './ProductBotones';


import { ProductImage } from './ProductImage'
import { ProductTitle } from './ProductTitle'
import { ProductBotons } from './ProductBotones'
import { ProductCard as ProductImageHoc } from './ProductCard'
import { ProductCardHocProps } from '../interfaces/product.inteface';



export const ProductCard:ProductCardHocProps = Object.assign(ProductImageHoc, {
    Image: ProductImage,
    Title: ProductTitle,
    Botons: ProductBotons,
});


export default ProductCard;