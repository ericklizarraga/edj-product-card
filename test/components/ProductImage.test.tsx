
import React from "react";

import render from "react-test-renderer";
import { ProductImage, ProductCard } from '../../src/components';
import { product2 } from "./data/products";


describe('ProductImage', () => {

    test('debe de mostrar el componente correctamente con la imagen perzonalizaado', () => {

        const wrapper = render.create(
            <ProductImage  img="./coffee-mug2.pn" />
        );

        expect(wrapper.toJSON()).toMatchSnapshot();

    });

    test('debe  de mostrar el componente con la imagen del  producto', () => {

        const wrapper = render.create(
            <ProductCard product={product2}>
                {
                    () => (
                        <ProductImage />
                    )
                }
            </ProductCard>
        );

        expect(wrapper.toJSON()).toMatchSnapshot();

    });
});