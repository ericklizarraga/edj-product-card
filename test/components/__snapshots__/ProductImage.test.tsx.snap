// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage debe  de mostrar el componente con la imagen del  producto 1`] = `null`;

exports[`ProductImage debe de mostrar el componente correctamente con la imagen perzonalizaado 1`] = `null`;
